include/sdf/parser.hh

```diff
@@ -333,6 +333,17 @@ namespace sdf
   SDFORMAT_VISIBLE
   bool convertFile(const std::string &_filename, const std::string &_version,
                    const ParserConfig &_config, SDFPtr _sdf);
+  /// \brief Convert an SDF file to a specific SDF version.
+  /// \param[in] _filename Name of the SDF file to convert.
+  /// \param[in] _version Version to convert _filename to.
+  /// \param[in] _config Custom parser configuration
+  /// \param[out] _sdf Pointer to the converted SDF document.
+  /// \param[out] _errors Vector of errors.
+  /// \return True on success.
+  SDFORMAT_VISIBLE
+  bool convertFile(const std::string &_filename, const std::string &_version,
+                   const ParserConfig &_config, SDFPtr _sdf,
+                   sdf::Errors &_errors);
```

Comment:
hot take:

for the new APIs, let's return `sdf::Errors` and pass the `SDFPtr` as the first argument and make ParserConfig optional

~~~
sdf::Errors convertFile(
    SDFPtr _sdf,
    const std::string &_filename,
    const std::string &_version,
    const ParserConfig &_config = ParserConfig::GlobalConfig())
~~~

Reply:
Wouldn't this leave it a bit weird in the sense that some of the overload functions return an `sdf::Errors` structure while others a `bool`? Also to add to the argument, the most common way of obtaining `sdf::Errors` in the library is through a parameter.

Reply:
all the `Load(sdf::ElementPtr)` functions return `Errors` as well as the `Resolve*` functions in the DOM classes. If a method already returns something like a `SDFPtr`, I think we need to pass the `Errors` by reference, but for APIs that would otherwise return `bool`, I think it makes more sense to return `Errors`

@azeey your thoughts?

Reply:
I had a deeper look at this suggestion and I found one more drawback: 

Because the old functions are now calling the new ones (i.e.: [here](https://github.com/gazebosim/sdformat/blob/ecb8fa225c1638095a28c1774d4598078f85d590/src/parser.cc#L1944)) it makes it very hard to figure out when to return `False` or `True` in these old functions if the new function is only returning `sdf::Errors`. This is due to the fact that some of the code down the line might add and `sdf::Error` and not return anything at that point (i.e.: [here](https://github.com/gazebosim/sdformat/blob/ecb8fa225c1638095a28c1774d4598078f85d590/src/Converter.cc#L171)). So I'm not sure if this change can be applied without breaking behavior or having some code duplication on the old functions...

Reply:
Since the `Errors` object is going to be passed to multiple functions, I think it's better to pass it as an output parameter. Otherwise, a function will need to aggregate `Errors` from all the functions it calls, which would add a lot of overhead.

Reply:
> Since the `Errors` object is going to be passed to multiple functions, I think it's better to pass it as an output parameter.

I think it depends on the context. There are methods in our API that pass Errors along but don't check them. If you don't care about the Errors, then it is simpler to pass the Errors object by reference. If you do care about the result, however, then it's a little bit more work because you have to create a clean `Errors` object to pass in and then possibly insert those errors back into the vector of accumulated errors.

Reply:
In the case where we do care about the Errors object, inserting the errors back into the vector of accumulated errors would happen regardless of whether we passed by reference or returned an `Errors` object. The only difference would be whether we have an extra line creating a new `Errors` object or not.

I think returning `Errors` is more convenient for APIs that are main entry points, such as `Root::Load`. You could argue `convertFile` is also a main entry point, but as @marcoag said, determining the boolean value of the returned `Errors` object is not straightforward for the functions that are being overloaded.

Reply:
my proposed API puts the `SDFPtr` as the first argument, which is different enough from the existing API that it should hopefully be obvious which overload is being used. Putting the output variables first allows optional `const` arguments at the end so we could reduce the number of overloads

Reply:
The problem is not knowing which overload is being used. It's that the old function (one that doesn't take `sdf::Error`) is now implemented in terms of the new function. But the old function returns a bool so if we make the new function return an `sdf::Error`, how would we interpret that to a bool.

Reply:
`sdf::Errors` is a vector, so return `true` if it is empty: `return errors.empty();`

Reply:
> some of the code down the line might add and sdf::Error and not return anything at that point (i.e.: [here](https://github.com/gazebosim/sdformat/blob/ecb8fa225c1638095a28c1774d4598078f85d590/src/Converter.cc#L171)).

If we return `false` with that error the behavior of that function and any other users of it will change.

Reply:
> some of the code down the line might add and sdf::Error and not return anything at that point (i.e.: [here](https://github.com/gazebosim/sdformat/blob/ecb8fa225c1638095a28c1774d4598078f85d590/src/Converter.cc#L171)).
> 
> If we return `false` with that error the behavior of that function and any other users of it will change.
> 
> In short: in the current behavior, not always there is an error in the `sdf::Errors` false is returned.

Your example is a different function `Converter::Convert`. I'm only proposing this change for the `convertFile` and `convertString` functions. I'm not sure how I understand how your comment applies to `convertFile` and `convertString`

Reply:
But `convertFile` calls `Converter::Convert` [here](https://github.com/gazebosim/sdformat/blob/5f91d45a201662332d045ab845cda5ddea827e96/src/parser.cc#L1992), so there's a chance that the `sdf::Errors` being passed there gets an error and no `false` is returned, isn't it?

Reply:
all the errors will be returned, so I don't think that would be a problem

here is what I would do for `convertFile`: dcbedd7654cccb055605578e7b7a10aad2c0b974

Reply:
> But `convertFile` calls `Converter::Convert` [here](https://github.com/gazebosim/sdformat/blob/5f91d45a201662332d045ab845cda5ddea827e96/src/parser.cc#L1992), so there's a chance that the `sdf::Errors` being passed there gets an error and no `false` is returned, isn't it?

an `sdf::Errors` vector is passed to `Converter::Convert`, but I believe `Convert` only returns `false` if it adds an error to the vector, so I think it will be fine with the changes I have proposed in #1188

Reply:
Merged proposal. Thanks!